### PR TITLE
Issue #208 improve performance of project imports with a lot of parents

### PR DESCRIPTION
**Pull request by caroso-de**

### CODE REVIEW

**Review by HannesWell — CHANGES_REQUESTED** [review]
First of all, thank you for this PR. I think in general this is a good improvement.

But before we can get into the technical part, I have to ask you to create an Eclipse account and electronically sign the Eclipse Contributor Agreement (ECA), as it is described at [m2e's Contributing guide](https://github.com/eclipse-m2e/m2e-core/blob/master/CONTRIBUTING.md).
This usually does not take too long, but it is mandatory.

Once you have signed the ECA you have to make sure that the author e-mail address of subsequent commits is the same as the one used in the ECA.
As soon this PR is ready to be merged, we can then squash all subsequent commits into one change and set the author E-Mail accordingly.

**Comment by caroso-de** [comment]
I'm happy you agree to the general solution. Having next to no in-depth knowledge of this project I wasn't sure I'm heading in the right direction with this up until now. 

> But before we can get into the technical part, I have to ask you to create an Eclipse account and electronically sign the Eclipse Contributor Agreement (ECA), as it is described at [m2e's Contributing guide](https://github.com/eclipse-m2e/m2e-core/blob/master/CONTRIBUTING.md).
> This usually does not take too long, but it is mandatory.

Ok, I'll see into this. And thanks for the review

File: org.eclipse.m2e.binaryproject.tests/src/org/eclipse/m2e/binaryproject/tests/BinaryProjectTest.java
Comment (caroso-de) [reply org.eclipse.m2e.binaryproject.tests/src/org/eclipse/m2e/binaryproject/tests/BinaryProjectTest.java:43]: I'd say most of them. I still have problems to understand why and when those tests fail on my machine. The only test that failed consistently was the MemoryConsumptionTest iirc.

But i agree, we should re-enable those tests before merging.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingFactory.java
Comment (caroso-de) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingFactory.java:878]: I'm sorry ths comment totally slipped my review. It's just a note I made for myself while thinking about possible solutions. Honestly, I don't see that it contributes much to readability. I'd rather remove it along with the `TODO` regarding reuse and add some documentation to the new lifecycle class.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryManager.java
Comment (caroso-de) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryManager.java:481]: You're right (obviously). Sorry for missing that

File: org.eclipse.m2e.discovery/META-INF/MANIFEST.MF
Comment (caroso-de) [reply org.eclipse.m2e.discovery/META-INF/MANIFEST.MF:5]: Agreed. Regarind Version Bumps: Is there an integrated/automated version bump, because I cannot remember ever having set the version to 1.18.2 myself.
Additionally: What do you think should the bumped m2e.core version be. Strictly speaking it should be 2.0.0, right?

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingContext.java
Comment (caroso-de) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingContext.java:38]: Agreed, the method works more like what you'd expect from a cache this way. 

However I'd usually refrain from naming something like this a "getter" but maybe this is a mater of personal taste. If you prefer "get", let's build it this way.

Please let me know.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingFactory.java
Comment (HannesWell) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingFactory.java:878]: Yes, that's fine.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingContext.java
Comment (HannesWell) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/lifecyclemapping/LifecycleMappingContext.java:38]: If you prefer 'determine' this is also fine. Another alternative could be 'compute' (which is for example used by the Map interface). I think in either case it is comprehensible and that is the important part.

What is more important to me is the content of the method and that both methods are merged.

File: org.eclipse.m2e.discovery/META-INF/MANIFEST.MF
Comment (HannesWell) [reply org.eclipse.m2e.discovery/META-INF/MANIFEST.MF:5]: > Agreed. Regarind Version Bumps: Is there an integrated/automated version bump, because I cannot remember ever having set the version to 1.18.2 myself.

Actually there is nothing installed through m2e's Oomph setup that can bump versions automatically and I don't know any plug-in/tool that bumps automatically. There are the PDE API-Tools that can check for required bumps, but they don't bump automatically and m2e is not configured for them.

> Additionally: What do you think should the bumped m2e.core version be. Strictly speaking it should be 2.0.0, right?

No need for that. All classes you have changed are internal (note the `internal` segment in the package name) and not part of the public API and therefore can be changed arbitrarily without further notice. So a micro/service version bump by one is sufficient.

Comment (HannesWell) [reply org.eclipse.m2e.discovery/META-INF/MANIFEST.MF:5]: The only thing that need to be coordinated are the tests in `m2e-core-tests`. These are actually in another repository and included as git submodule. So the changes to them have to be addressed to another repository (https://github.com/tesla/m2e-core-tests). This structure was grown historically.

But we have a chicken-egg-problem here, since the test fail if you change the method-signature here but not in the tests and you cannot change the tests first because they would fail with compile errors because they are compiled against the current methods.
So either: 
- you first only add corresponding method overloads with additional context argument and keep the old methods delegating to the new ones and passing null as context (then perform a null check before using the context, but you should to this anyway), then update the tests to use the new methods with another PR to the mentioned repo and once the second PR is merged, create a third PR to remove the old methods.
- or parallel to this PR the PR for the tests repo is created and once both are ready to be merged (and we are sure everything will work fine) we merge them immediately one after the other and ignore that builds fail for a short time.

@mickaelistria do you have a preferred way how this is handled?

File: org.eclipse.m2e.binaryproject.tests/src/org/eclipse/m2e/binaryproject/tests/BinaryProjectTest.java
Comment (mickaelistria) [reply org.eclipse.m2e.binaryproject.tests/src/org/eclipse/m2e/binaryproject/tests/BinaryProjectTest.java:43]: We should basically never ignore them ;)

**Review by mickaelistria — CHANGES_REQUESTED** [review]
I won;t review until some existing tests are ignored or failing. Working tests are a prerequisite for reviw to be interesting, tere is no point in reviewing code whicn in th end may have some regression.

**Comment by caroso-de** [comment]
> I won;t review until some existing tests are ignored or failing. Working tests are a prerequisite for reviw to be interesting, tere is no point in reviewing code whicn in th end may have some regression.

I agree regarding not ignoring the tests (I'll update the pr accordingly). However I do need help with some of the tests because with my machine/setup, they fail on plain master, too.

**Comment by HannesWell** [comment]
> I agree regarding not ignoring the tests (I'll update the pr accordingly). However I do need help with some of the tests because with my machine/setup, they fail on plain master, too.
> 
> edit: and there's also the issue of m2e-core-tests

At the moment we are in the process of fixing tests. Please see issue #271 for this. I have already fixed some tests that failed only on windows and we are waiting for the release of another library in order to fix the test that also fails during build.

So please make sure you have updated to the current master for m2e-core as well as for the m2e-core-tests submodule! You can merge the current master into your `caroso/issue-208-perf-2` branch for this.

After this and after you have updated the m2e-core-tests submodule which test keep failing and what's the cause?

Please also not that, because you probably have to modify internal public methods, the tests in the submodule also have to be changed. Therefore it would be great, that once we have completed the review of this PR that you provide another PR to the https://github.com/tesla/m2e-core-tests repo with the adjusted tests. We can then test this together and submit it closely to each other.

**Comment by caroso-de** [comment]
Just an update: I'd still very much like to finish this pr, but due to unforseen circumstances I'm unable to work on this for the next few weeks.

**Comment by mickaelistria** [comment]
Related bug was resolved, let's close this PR.

**Pull request closed without merging**
